cpp/helloworld/greeter_async_server.cc

```diff
@@ -0,0 +1,134 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <memory>
+#include <iostream>
+#include <string>
+#include <thread>
+
+#include <grpc/grpc.h>
+#include <grpc/support/log.h>
+#include <grpc++/async_unary_call.h>
+#include <grpc++/completion_queue.h>
+#include <grpc++/server.h>
+#include <grpc++/server_builder.h>
+#include <grpc++/server_context.h>
+#include <grpc++/server_credentials.h>
+#include <grpc++/status.h>
+#include "helloworld.grpc.pb.h"
+
+using grpc::CompletionQueue;
+using grpc::Server;
+using grpc::ServerAsyncResponseWriter;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::Status;
+using helloworld::HelloRequest;
+using helloworld::HelloReply;
+using helloworld::Greeter;
+
+static bool got_sigint = false;
+
+class ServerImpl final {
+ public:
+  ServerImpl() : service_(&service_cq_) {}
+
+  ~ServerImpl() {
+    server_->Shutdown();
+    rpc_cq_.Shutdown();
+    service_cq_.Shutdown();
+  }
+
+  // There is no shutdown handling in this code.
+  void Run() {
+    std::string server_address("0.0.0.0:50051");
+
+    ServerBuilder builder;
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+    builder.RegisterAsyncService(&service_);
+    server_ = builder.BuildAndStart();
+    std::cout << "Server listening on " << server_address << std::endl;
+
+    while (true) {
+      CallData* rpc = new CallData();
+      service_.RequestSayHello(&rpc->ctx, &rpc->request, &rpc->responder,
+                               &rpc_cq_, rpc);
+      void* got_tag;
+      bool ok;
+      service_cq_.Next(&got_tag, &ok);
+      GPR_ASSERT(ok);
+      GPR_ASSERT(got_tag == rpc);
+
+      std::thread t(&ServerImpl::HandleRpc, this, rpc);
+      t.detach();
+    }
+  }
+
+ private:
+  struct CallData {
+    CallData() : responder(&ctx) {}
+    ServerContext ctx;
+    HelloRequest request;
+    HelloReply reply;
+    ServerAsyncResponseWriter<HelloReply> responder;
+  };
+
+  // Runs in a detached thread, processes rpc then deletes data.
+  void HandleRpc(CallData* rpc) {
+    std::string prefix("Hello ");
+    rpc->reply.set_message(prefix + rpc->request.name());
+    rpc->responder.Finish(rpc->reply, Status::OK, &rpc->ctx);
+    void* got_tag;
+    bool ok;
+    rpc_cq_.Next(&got_tag, &ok);
```

Comment:
This will only handle one outstanding RPC at a time.

Suggest:
- making server_cq_, rpc_cq_ the same completion queue
- having a state variable on CallData
- having a virtual function on CallData to increment the state

your main loop will be:
```c
while (true) {
  service_cq_.Next(&tag, &ok);
  static_cast<CallData*>(tag)->OnComplete();
}
```